java/lib/src/test/java/frequency/LibraryTest.java

```diff
@@ -6,13 +6,133 @@
+    @Test
+    void logger_double_initialize_should_fail() {
+        Native.loggerInitialize(Level.DEBUG.toInt(), logger);
+        assertEquals(true, testLogsForPattern(Level.WARN, "Duplicate logger initialization ignored"));
+    }
+
+    @Test
+    void logger_debug_should_log() {
+        log(Level.DEBUG.toInt(), "This is a debug log");
+        log(Level.INFO.toInt(), "This is an info log");
+        log(Level.WARN.toInt(), "This is a warning log");
+        log(Level.ERROR.toInt(), "This is an error log");
+        assertEquals(true, testLogsForPattern(Level.DEBUG, "This is a debug log"));
+        assertEquals(true, testLogsForPattern(Level.INFO, "This is an info log"));
+        assertEquals(true, testLogsForPattern(Level.WARN, "This is a warning log"));
+        assertEquals(true, testLogsForPattern(Level.ERROR, "This is an error log"));
+    }
+
+    @Test
+    void logger_info_should_log() {
+        Native.loggerSetMaxLevel(Level.INFO.toInt());
+        log(Level.DEBUG.toInt(), "This is a debug log");
+        log(Level.INFO.toInt(), "This is an info log");
+        log(Level.WARN.toInt(), "This is a warning log");
+        log(Level.ERROR.toInt(), "This is an error log");
+        assertEquals(false, testLogsForPattern(Level.DEBUG, "This is a debug log"));
+        assertEquals(true, testLogsForPattern(Level.INFO, "This is an info log"));
+        assertEquals(true, testLogsForPattern(Level.WARN, "This is a warning log"));
+        assertEquals(true, testLogsForPattern(Level.ERROR, "This is an error log"));
+    }
+
+    @Test
+    void logger_warn_should_log() {
+        Native.loggerSetMaxLevel(Level.WARN.toInt());
+        log(Level.DEBUG.toInt(), "This is a debug log");
+        log(Level.INFO.toInt(), "This is an info log");
+        log(Level.WARN.toInt(), "This is a warning log");
+        log(Level.ERROR.toInt(), "This is an error log");
+        assertEquals(false, testLogsForPattern(Level.DEBUG, "This is a debug log"));
+        assertEquals(false, testLogsForPattern(Level.INFO, "This is an info log"));
+        assertEquals(true, testLogsForPattern(Level.WARN, "This is a warning log"));
+        assertEquals(true, testLogsForPattern(Level.ERROR, "This is an error log"));
+    }
+
+    @Test
+    void logger_error_should_log() {
+        Native.loggerSetMaxLevel(Level.ERROR.toInt());
+        log(Level.DEBUG.toInt(), "This is a debug log");
+        log(Level.INFO.toInt(), "This is an info log");
+        log(Level.WARN.toInt(), "This is a warning log");
+        log(Level.ERROR.toInt(), "This is an error log");
+        assertEquals(false, testLogsForPattern(Level.DEBUG, "This is a debug log"));
+        assertEquals(false, testLogsForPattern(Level.INFO, "This is an info log"));
+        assertEquals(false, testLogsForPattern(Level.WARN, "This is a warning log"));
+        assertEquals(true, testLogsForPattern(Level.ERROR, "This is an error log"));
+    }
+
+    private static native void log(int level, String message);
```

Comment:
Should we keep all native functions inside Native even if it's for tests?

Reply:
I initially tried that, but Java doesn't have conditional compilation, and I didn't want to pollute the lib/jar in production code with functions that are only for test. I already had to do that on the Rust side, but I don't want to propagate that. At least the Rust function is technically valid for production (though unnecessary)